### ISSUE TITLE
Compiling project creates an out-of-memory error

### ISSUE THREAD

**Issue by pharpend** [yesodweb/yesod#998]
# Specifications

Full tree: <https://github.com/pharpend/snowdrift>. 

These are the situations in which I can reproduce this:

Operating system | GHC version | yesod-core version | File on which GHC runs out of memory | Memory on system
-----------------|-------------|--------------------|--------------------------------------|-----------------
Arch Linux       | 7.10.1      | `==1.4.9.*`        | Model.hs                             | 16GB
FreeBSD          | 7.8.4       | `==1.4.9.*`        | Foundation.hs                        | 1GB
FreeBSD          | 7.8.4       | `==1.4.8.*`        | Foundation.hs                        | 1GB
FreeBSD          | 7.8.4       | `<1.4.8`           | Foundation.hs                        | 1GB

Note that "yesod-core version" means "when I have this version of yesod-core installed, this behavior occurs".

# Steps to reproduce:

```
git clone git://github.com/pharpend/snowdrift.git
cd snowdrift
cabal sandbox init
cabal install
```

Turning off optimizations (with `-fdev`) fixes the issue, or (on Arch Linux), switching down to `==1.4.8.*`. (See the table above).

**Comment by gregwebs** [comment]
Thanks for the report! Is there a difference between `-O2` and using just `-O`? I am planning on changing the scaffolding to to use just `-O` (which is recommended by the user guide)

**Comment by pharpend** [comment]
> Thanks for the report! Is there a difference between `-O2` and using
> just `-O`? I am planning on changing the scaffolding to to use just
> `-O` (which is recommended by the user guide)

Nope, still fails. I only have access to the Arch box at the immediate
moment (mainly I'm tired and too lazy to go to the other room and turn
the BSD box on), but I would imagine it's the same result.

Thanks,

Peter Harpending

**Comment by gregwebs** [comment]
In general I just expect to need 2 GB of memory to compile a medium-sized Haskell program quickly. So the failures with 1 GB are not entirely unexpected and we should focus on the failure with 16 GB.

**Comment by gregwebs** [comment]
On a hunch, I remembered we had an issue with horrible compilation behavior of the function [`toCaseFold` in text](https://github.com/bos/text/issues/116). I searched the persistent repo for this function and found a usage in the migration code. I removed the `mkMigrate` reference from Model.hs, and then the code compiled :)

**Comment by pharpend** [comment]
On Wed, May 27, 2015 at 08:11:03PM -0700, Greg Weber wrote:
> I removed the `mkMigrate` reference from Model.hs, and then the code
> compiled :)

I'm not entirely sure what you mean here. Could you post a diff?

**Comment by pharpend** [comment]
If I try this:

```diff
diff --git a/Model.hs b/Model.hs
index 1ce4b47..ac9fb9c 100644
--- a/Model.hs
+++ b/Model.hs
@@ -46,7 +46,6 @@ import Yesod.Markdown              (Markdown)
 -- at:
 -- http://www.yesodweb.com/book/persistent/
 share [ mkPersist sqlSettings
-      , mkMigrate "migrateAll"
       , mkDeleteCascade sqlSettings
       , mkDiscussionTypes
       , mkReferences "Comment"
```

I get this string of errors:

```
Resolving dependencies...
In order, the following will be installed:
Snowdrift-0.1.4 (reinstall) (changes: cmdargs-0.10.13 added, email-validate-2.1.2 added, mime-0.4.0.2 added, mwc-random-0.13.3.2 added, yesod-1.4.1.5 added, yesod-auth-1.4.5 added, yesod-auth-hashdb-1.4.2.1 added, yesod-core-1.4.8.3 -> 1.4.9.1, yesod-form-1.4.4.1 added, yesod-markdown-0.9.4 added, yesod-newsfeed-1.4.0.1 added, yesod-static-1.4.0.4 added)
Warning: Note that reinstalls are always dangerous. Continuing anyway...
Notice: installing into a sandbox located at
/home/pete/src/snowdrift/.cabal-sandbox
Configuring Snowdrift-0.1.4...
Building Snowdrift-0.1.4...
Failed to install Snowdrift-0.1.4
Build log ( /home/pete/src/snowdrift/.cabal-sandbox/logs/Snowdrift-0.1.4.log ):
Configuring Snowdrift-0.1.4...
Building Snowdrift-0.1.4...
Preprocessing library Snowdrift-0.1.4...
[ 83 of 104] Compiling Model.Project    ( Model/Project.hs, dist/dist-sandbox-7eab385b/build/Model/Project.o )
[ 84 of 104] Compiling Handler.Widget   ( Handler/Widget.hs, dist/dist-sandbox-7eab385b/build/Handler/Widget.o ) [Foundation changed]
[ 85 of 104] Compiling Model.Wiki       ( Model/Wiki.hs, dist/dist-sandbox-7eab385b/build/Model/Wiki.o ) [Foundation changed]
[ 86 of 104] Compiling SnowdriftEventHandler ( SnowdriftEventHandler.hs, dist/dist-sandbox-7eab385b/build/SnowdriftEventHandler.o ) [Foundation changed]
[ 87 of 104] Compiling Widgets.Time     ( Widgets/Time.hs, dist/dist-sandbox-7eab385b/build/Widgets/Time.o )
[ 88 of 104] Compiling Handler.Image    ( Handler/Image.hs, dist/dist-sandbox-7eab385b/build/Handler/Image.o )
[ 89 of 104] Compiling Handler.Notification ( Handler/Notification.hs, dist/dist-sandbox-7eab385b/build/Handler/Notification.o ) [Foundation changed]
[ 90 of 104] Compiling Widgets.ProjectPledges ( Widgets/ProjectPledges.hs, dist/dist-sandbox-7eab385b/build/Widgets/ProjectPledges.o ) [Foundation changed]
[ 91 of 104] Compiling View.User        ( View/User.hs, dist/dist-sandbox-7eab385b/build/View/User.o ) [Foundation changed]
[ 92 of 104] Compiling View.Project     ( View/Project.hs, dist/dist-sandbox-7eab385b/build/View/Project.o ) [Data.Filter changed]
[ 93 of 104] Compiling Model.Blog       ( Model/Blog.hs, dist/dist-sandbox-7eab385b/build/Model/Blog.o ) [Foundation changed]
[ 94 of 104] Compiling View.Comment     ( View/Comment.hs, dist/dist-sandbox-7eab385b/build/View/Comment.o ) [Foundation changed]
[ 95 of 104] Compiling Handler.Comment  ( Handler/Comment.hs, dist/dist-sandbox-7eab385b/build/Handler/Comment.o ) [Foundation changed]
[ 96 of 104] Compiling Handler.SnowdriftEvent ( Handler/SnowdriftEvent.hs, dist/dist-sandbox-7eab385b/build/Handler/SnowdriftEvent.o ) [Foundation changed]
[ 97 of 104] Compiling Handler.User.Comment ( Handler/User/Comment.hs, dist/dist-sandbox-7eab385b/build/Handler/User/Comment.o ) [Foundation changed]
[ 98 of 104] Compiling Handler.ProjectBlog ( Handler/ProjectBlog.hs, dist/dist-sandbox-7eab385b/build/Handler/ProjectBlog.o ) [Foundation changed]
[ 99 of 104] Compiling Handler.User     ( Handler/User.hs, dist/dist-sandbox-7eab385b/build/Handler/User.o ) [Foundation changed]
[100 of 104] Compiling View.SnowdriftEvent ( View/SnowdriftEvent.hs, dist/dist-sandbox-7eab385b/build/View/SnowdriftEvent.o ) [Foundation changed]
[101 of 104] Compiling Handler.Project  ( Handler/Project.hs, dist/dist-sandbox-7eab385b/build/Handler/Project.o )
[102 of 104] Compiling Handler.Wiki.Comment ( Handler/Wiki/Comment.hs, dist/dist-sandbox-7eab385b/build/Handler/Wiki/Comment.o ) [Foundation changed]
[103 of 104] Compiling Handler.Wiki     ( Handler/Wiki.hs, dist/dist-sandbox-7eab385b/build/Handler/Wiki.o ) [Foundation changed]
[104 of 104] Compiling Application      ( Application.hs, dist/dist-sandbox-7eab385b/build/Application.o ) [Foundation changed]

Application.hs:172:48:
    Not in scope: ‘migrateAll’
    Perhaps you meant ‘migrate’ (imported from Import)
cabal: Error: some packages failed to install:
Snowdrift-0.1.4 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by snoyberg** [comment]
There's only one error, and to fix it, you'd need to remove the usage of `migrateAll` in Application.hs.

**Comment by pharpend** [comment]
@snoyberg Tried that, it seems to compile okay. (I should probably get in the habit of actually reading error messages).

I have this diff, which seems to compile:

```diff
diff --git a/Application.hs b/Application.hs
index 53fd686..2a97a01 100644
--- a/Application.hs
+++ b/Application.hs
@@ -168,9 +168,12 @@ makeFoundation conf = do
                         liftIO $ putStrLn "ready."
         _ -> return ()
 
-    let migration = runSqlPool
-            (doManualMigration >> runMigration migrateAll >> migrateTriggers)
-            pool
+    let migration = 
+          flip runSqlPool
+               pool
+               (do doManualMigration 
+                   -- runMigration migrateAll 
+                   migrateTriggers)
 
     void $ runLoggingT
         migration
@@ -244,7 +247,7 @@ deprecatedApplyManualMigrations = do
 -- storing the necessary sql statements to commit and share.
 saveUnsafeMigrations :: (MonadIO m, Functor m) => ReaderT SqlBackend m ()
 saveUnsafeMigrations = do
-    unsafe <- (L.filter fst) <$> parseMigration' migrateAll
+    unsafe <- pure mempty
     unless (L.null $ L.map snd unsafe) $ do
         liftIO $ T.writeFile filename $ T.unlines $ L.map ((`snoc` ';') . snd) unsafe
         liftIO $ mapM_ (T.hPutStrLn stderr) unsafeMigMessages
diff --git a/Model.hs b/Model.hs
index 1ce4b47..0a7b2c6 100644
--- a/Model.hs
+++ b/Model.hs
@@ -46,7 +46,7 @@ import Yesod.Markdown              (Markdown)
 -- at:
 -- http://www.yesodweb.com/book/persistent/
 share [ mkPersist sqlSettings
-      , mkMigrate "migrateAll"
+      -- , mkMigrate "migrateAll"
       , mkDeleteCascade sqlSettings
       , mkDiscussionTypes
       , mkReferences "Comment"
```

**Comment by gregwebs** [comment]
Tthe goal is just to get Model.hs to compile, the full diff you have is what you would need though to follow through for compilation of the whole project.

Of course, you would probably like to have your migration code back!
The `toCaseFold` issue was fixed in text 1.2.0.5.

I removed the changes to the code and updated to text to 1.2.1.1 and unfortunately Model.hs still does not compile, so there must be something else going on there. Still, you should definitely upgrade your version of `text` for the project to `>= 1.2.0.5`.

I am pretty happy that I narrowed the issue down as much as I have already and am not planning on digging in further in the near future.

**Comment by gregwebs** [comment]
I am moving this issue over to the persistent repo: https://github.com/yesodweb/persistent/issues/403
